psi4/driver/procrouting/mcscf/mcscf_solver.py

```diff
@@ -44,6 +44,53 @@ def print_iteration(mtype, niter, energy, de, orb_rms, ci_rms, nci, norb, stype)
     core.print_out("%s %2d:  % 18.12f   % 1.4e  %1.2e  %1.2e  %3d  %3d  %s\n" %
                     (mtype, niter, energy, de, orb_rms, ci_rms, nci, norb, stype))
 
+def build_semicanonical_mos(ciwfn):
+
+    # Grab Fock matrices and build the average Fock operator
+    mcscf_obj = ciwfn.mcscf_object()
+    afock = mcscf_obj.current_AFock()
+    ifock = mcscf_obj.current_IFock()
+    Favg = afock.clone()
+    Favg.add(ifock)
+
+    # Grab orbital dimensions
+    nirrep = Favg.nirrep()
+    offsets = [0 for i in range(nirrep)]
+    nmopi = Favg.coldim()
+    noccpi = mcscf_obj.noccpi()
+    nactpi = mcscf_obj.nactpi()
+    nvirpi = mcscf_obj.nvirpi()
+
+    U = core.Matrix("U to semi", nmopi, nmopi)
+
+    # Diagonalize each block of Favg
+    for block in [noccpi,nactpi,nvirpi]:
+        F = core.Matrix("Fock",block,block)
+        for h in xrange(nirrep):
+            offset = offsets[h]
+            for i in xrange(block[h]):
+                for j in xrange(block[h]):
+                    F.set(h, i, j, Favg.get(h, i + offset, j + offset))
```

Comment:
You could consider the numpy slicing notation here:

```
for h in range(nirrep):
     offset = offsets[h]
     sl = slice(offset, offset + block[h])
     F.nph[h][:] = Favg.nph[h, sl, sl]
```

The same thing can be done with U below. This will also be much faster. Last comment: `xrange` is no longer valid with Python3, probably best to drop that.